File: Python version/languages/pddl/Plan.py
from .action import Action
from base.output import Output
from languages.pddl.pddl_mapper import PDDLMapper
from abc import ABCMeta


class Plan(Output):
    """A simplified solution to a PDDL problem."""
    __metaclass__ = ABCMeta

    def __init__(self, plan, error):
        super(Plan, self).__init__(plan, error)
        self._action_sequence = None
        self.__actions_objects = None

    def get_actions(self):
        """Returns a set of Actions."""
        if self._action_sequence is None:
            self._action_sequence = list()
            self._parse()
        return self._action_sequence

    def get_actions_objects(self):
        """Returns a set of Objects representing Actions."""
        if self.__actions_objects is None:
            self.__actions_objects = list()
            mapper = PDDLMapper.get_instance()
            for a in self.get_actions():
                obj = mapper.get_object(a.get_name())
                if obj is not None:
                    self.__actions_objects.append(obj)
        return self.__actions_objects

    def store_action(self, action):
        self._action_sequence.append(Action(action))

File: Java version/app/src/main/java/it/unical/mat/embasp/languages/pddl/Plan.java
package it.unical.mat.embasp.languages.pddl;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import it.unical.mat.embasp.base.Output;
import it.unical.mat.parsers.pddl.PDDLDataCollection;

/** A simplified solution to a PDDL problem */

public abstract class Plan extends Output implements PDDLDataCollection {
	protected List<Action> actionSequence;

	private List<Object> actionsObjects;

	public Plan(final String plan, final String error) {
		super(plan, error);
	}

	public List<Action> getActions() {
		if (actionSequence == null) {
			actionSequence = new ArrayList<>();
			parse();
		}

		return Collections.unmodifiableList(actionSequence);
	}

	public List<Object> getActionsObjects() throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
		if(actionsObjects == null) {
			actionsObjects = new ArrayList <> ();

			for(final Action action : getActions()) {
				final Object obj = PDDLMapper.getInstance().getObject(action.getName());
				
				if (obj != null)
					actionsObjects.add(obj);
			}
		}

		return actionsObjects;
	}
	
	@Override
	public void storeAction(final String action) {
		actionSequence.add(new Action(action));
	}
}
